macros/src/str.rs

```diff
@@ -0,0 +1,376 @@
+use std::fmt::{Display, Formatter};
+
+const MAX_I128_REPR: i128 = 0x0000_0000_FFFF_FFFF_FFFF_FFFF_FFFF_FFFF;
+
+#[derive(Debug)]
+pub struct Unpacked {
+    pub mantissa: i128,
+    pub scale: u32,
+}
+
+impl Unpacked {
+    pub const fn lo(&self) -> u32 {
+        self.mantissa.unsigned_abs() as u32
+    }
+
+    pub const fn mid(&self) -> u32 {
+        (self.mantissa.unsigned_abs() >> 32) as u32
+    }
+
+    pub const fn hi(&self) -> u32 {
+        (self.mantissa.unsigned_abs() >> 64) as u32
+    }
+
+    pub const fn negative(&self) -> bool {
+        self.mantissa < 0
+    }
+}
+
+pub type ParseResult<'str> = Result<Unpacked, ParseError<'str>>;
+
+#[derive(Debug, PartialEq)]
+pub enum ParseError<'src> {
+    Empty,
+    ExceedsMaximumPossibleValue,
+    FractionEmpty,
+    InvalidExp(i32),
+    InvalidRadix(u32),
+    LessThanMinimumPossibleValue,
+    Underflow,
+    Unparseable(&'src [u8]),
+}
+
+impl Display for ParseError<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ParseError::Empty => write!(f, "Number is empty, must have an integer part."),
+            ParseError::ExceedsMaximumPossibleValue => {
+                write!(f, "Number exceeds maximum value that can be represented.")
+            }
+            ParseError::FractionEmpty => write!(f, "Fraction empty, consider adding a `0` after the period."),
+            ParseError::InvalidExp(exp) if *exp < 0 => write!(
+                f,
+                "Scale exceeds the maximum precision allowed: {} > {}.",
+                exp.unsigned_abs(),
+                28
+            ),
+            ParseError::InvalidExp(exp) => {
+                write!(f, "Invalid exp {exp} -- exp must be in the range -28 to 28 inclusive.")
+            }
+            ParseError::InvalidRadix(radix) => write!(
+                f,
+                "Invalid radix {radix} -- radix must be in the range 2 to 36 inclusive."
+            ),
+            ParseError::LessThanMinimumPossibleValue => {
+                write!(f, "Number less than minimum value that can be represented.")
+            }
+            ParseError::Underflow => write!(f, "Number has a high precision that can not be represented."),
+            ParseError::Unparseable(src) => {
+                write!(
+                    f,
+                    "Cannot parse decimal, unexpected \"{}\".",
+                    core::str::from_utf8(src).unwrap()
+                )
+            }
+        }
+    }
+}
+
+pub const fn parse_decimal(src: &str, exp: i32) -> ParseResult {
```

Comment:
I think having `exp` as a mandatory argument is incorrect since it can get overwritten in `parse_10` cases.